### PR TITLE
Dynamic enforcement improvements (4.0)

### PR DESCRIPTION
**Pull request by rjmccall**
These changes implement a series of improvements to the dynamic-enforcement runtime.  Specifically:

- It causes the runtime to remember a PC associated with both accesses and to print that PC when a conflict is detected.
- It changes the compiler and runtime to only emit a warning (i.e. to print to the console) when the source was compiled for Swift 3 compatibility.

This is an enhancement to the runtime for dynamic enforcement which requires minor code-generation changes associated with the calls to that runtime.  It does not affect code which is not being compiled for dynamic enforcement.

Fixes rdar://32127228.

This PR is low risk because the changes are straightforward and targeted.  The most conceptually complex part of this patch is collecting a meaningful PC for the access, and the only thing done with this information is printing it as an opaque pointer.

This PR has been tested with regression testing which checks for the new test output (to the extent possible without testing for specific pointer values), and with "eye testing" that the output is meaningful.

There will be a follow-up PR which improves on the diagnostic quality here; this PR is necessary in order to provide the information which that PR will use, as well as to implement the Swift 3 compatibility design.

### CODE REVIEW

**Comment by rjmccall** [comment]
@gparker42 Please review for Swift 4.0.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@atrick You might also be able to review this.

File: stdlib/public/runtime/Exclusivity.cpp [hunk 1]
```diff
@@ -61,8 +62,37 @@ static const char *getAccessName(ExclusivityFlags flags) {
   switch (flags) {
```
Comment (atrick) [inline stdlib/public/runtime/Exclusivity.cpp]: For clarity why not either call `getAccessAction` or assert that `flags == getAccessAction()`?

**Review by atrick — APPROVED** [review]
LGTM for swift4. One minor inlined comment could be addressed now or later.

File: stdlib/public/runtime/Exclusivity.cpp (hunk 1 above)
Comment (rjmccall) [inline stdlib/public/runtime/Exclusivity.cpp]: The assertion is a good idea, I'll do that on master.

**Pull request merged**